src/fcmp_pp/curve_trees.h

```diff
@@ -211,6 +211,108 @@ struct PathIndexes final
     std::vector<Range> layers;
 };
 
+// "Assigned" means the leaf is a member of the tree
+struct AssignedLeafIdx final
+{
+    bool assigned_leaf_idx{false};
+    uint64_t leaf_idx{0};
+
+    void assign_leaf(const uint64_t idx) { leaf_idx = idx; assigned_leaf_idx = true; }
+    void unassign_leaf() { leaf_idx = 0; assigned_leaf_idx = false; }
+
+    BEGIN_SERIALIZE_OBJECT()
+        FIELD(assigned_leaf_idx)
+        FIELD(leaf_idx)
+    END_SERIALIZE()
+
+    BEGIN_KV_SERIALIZE_MAP()
+        KV_SERIALIZE(assigned_leaf_idx)
+        KV_SERIALIZE(leaf_idx)
+    END_KV_SERIALIZE_MAP()
+};
+
+// Contains minimum path elems necessary for multiple paths (e.g. only contains the root once)
+struct ConsolidatedPaths final
+{
+    // Useful types to use in the code
+    std::unordered_map<uint64_t, std::vector<OutputContext>> leaves_by_chunk_idx;
+    std::vector<std::unordered_map<uint64_t, ChunkBytes>> layer_chunks_by_chunk_idx;
+
+    // Serializable helper types
+    struct LeafChunk
+    {
+        uint64_t chunk_idx;
+        std::vector<OutputContext> chunk;
+
+        BEGIN_KV_SERIALIZE_MAP()
+            KV_SERIALIZE(chunk_idx)
```

Comment:
These field names will be duplicated in the payload a lot, is that accept able for a consolidated format?